### PR TITLE
Custom diagnostic for lifetimes in method calls

### LINKED ISSUE #48899: Add a new parse error when encountering lifetimes in inappropriate places
[The following code](https://play.rust-lang.org/?gist=252a4b0cfca62aa5aa272afd013b11dc&version=nightly) should complain about the lifetime passed to `foo`:
```
fn foo<'a>(bar: &'a usize) {}

fn main() {
    let bar = 3;
    foo(&'a bar);
}
```

instead, it complains about `bar`:

```
error: expected `:`, found `bar`
 --> src/main.rs:5:13
  |
5 |     foo(&'a bar);
  |             ^^^ expected `:`
```

### PR DESCRIPTION
**Pull request by estebank**
Fix #48899.

### CODE REVIEW

File: src/test/ui/lifetimes/lifetime-in-method-call.rs [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+
+fn foo(x: &usize) {}
+
+fn main() {
+    'l: loop {
+        break 'l;
+    }
+    'y: loop {
+        let x = 3;
+        foo(&'a x);
```
Comment (nikomatsakis) [inline src/test/ui/lifetimes/lifetime-in-method-call.rs]: I'm confused -- what does this have to do with *method calls* in particular?

Comment (nikomatsakis) [reply src/test/ui/lifetimes/lifetime-in-method-call.rs]: That is, what happens if we do `let x = &'a y`?

File: src/libsyntax/parse/attr.rs [hunk 2]
```diff
@@ -59,6 +63,20 @@ impl<'a> Parser<'a> {
                     self.bump();
                     just_parsed_doc_comment = true;
                 }
+                token::Lifetime(l)
```
Comment (nikomatsakis) [inline src/libsyntax/parse/attr.rs]: I feel sort of surprised to see this error message in *this particular* function. It feels like this could be something that we (optionally) run after invocations to this function, perhaps? Or maybe a wrapper? Why is this function even called in these cases, I guess because of attributes on expressions?

Put another way, it just seems surprising to skip a lifetime *here* since it's not like it's part of a malformed outer attribute.

File: src/libsyntax/parse/parser.rs [hunk 3]
```diff
@@ -4284,7 +4284,7 @@ impl<'a> Parser<'a> {
                                    -> PResult<'a, Option<Stmt>> {
         maybe_whole!(self, NtStmt, |x| Some(x));
 
-        let attrs = self.parse_outer_attributes()?;
+        let attrs = self.parse_outer_attributes(true)?;
```
Comment (nikomatsakis) [inline src/libsyntax/parse/parser.rs]: why are we passing `true` here?

**Review by nikomatsakis — COMMENTED** [review]
This patch left me a bit confused.

File: src/test/ui/lifetimes/lifetime-in-method-call.rs (hunk 1 above)
Comment (estebank) [reply src/test/ui/lifetimes/lifetime-in-method-call.rs]: I called out method calls in particular because that was the mistake I saw in the wild, but yes, it should apply to anywhere that is not an ADT/method definition or a binding type.

From last nights quick test (as I was intrigued by the same):

```
error: found unexpected lifetime `'x`
 --> file.rs:2:25
  |
2 |     let x: &'x usize = &'x y;
  |                         ^^ unexpected lifetime
  |
  = help: lifetimes are only appropriate in definitions
```

File: src/libsyntax/parse/parser.rs (hunk 3 above)
Comment (estebank) [reply src/libsyntax/parse/parser.rs]: I have to confirm, but IIRC it is because otherwise `'label: loop {}` would have been rejected (before I added an explicit check for a `:`).

File: src/libsyntax/parse/attr.rs (hunk 2 above)
Comment (estebank) [reply src/libsyntax/parse/attr.rs]: Fair enough. I will rework the code to add a new method that proactively looks for lifetimes everywhere that applies.

**Comment by estebank** [comment]
I'm not entirely happy with the new change. I added a new `Parser` wide flag to determine wether a lifetime would be accepted or not in order to avoid adding a flag to almost every `parse_`-method, but this has the potential to be hard to maintain... What do you think, @nikomatsakis?

File: src/libsyntax/parse/parser.rs [hunk 4]
```diff
@@ -3823,13 +3840,9 @@ impl<'a> Parser<'a> {
...
                 self.expect_and()?;
+                self.maybe_disallow_lifetime();
                 let mutbl = self.parse_mutability();
-                if let token::Lifetime(ident) = self.token {
-                    let mut err = self.fatal(&format!("unexpected lifetime `{}` in pattern",
-                                                      ident));
-                    err.span_label(self.span, "unexpected lifetime");
-                    return Err(err);
-                }
+                self.maybe_disallow_lifetime();
```
Comment (estebank) [inline src/libsyntax/parse/parser.rs]: The only reason I added this second call is because the current check comes _after_ the `mut` parse, which is a weird place to have it, but I'm guessing not that uncommon of a mistake to write `foo(&mut 'a x)`...

File: src/libsyntax/parse/parser.rs [hunk 5]
```diff
@@ -245,6 +245,7 @@ pub struct Parser<'a> {
     pub desugar_doc_comments: bool,
     /// Whether we should configure out of line modules as we parse.
     pub cfg_mods: bool,
+    allow_lifetimes: bool,
```
Comment (nikomatsakis) [inline src/libsyntax/parse/parser.rs]: yeah this doesn't feel very good

Comment (nikomatsakis) [reply src/libsyntax/parse/parser.rs]: it seems a bit confusing to me though -- we are specifically looking for lifetimes that come after a `&`, right? i.e., that's the only place a user might expect to have them. Can we just add a check to the parsing of `&` in an expression context for lifetimes? (And forego the rest?)

**Comment by pietroalbini** [comment]
@estebank ping from triage! The reviewer left a comment that should be addressed.

**Comment by estebank** [comment]
@pietroalbini I need to rework the code, will be laptopless for two weeks. Feel free to close to clear up the queue.
